Update kyverno-policies chart with latest pod-security policies

treydock opened the pull request:
## Related issue

<!--
Please link the GitHub issue this pull request resolves in the format of `#1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->
Fixes #3063
Fixes #2277
Replaces PR #3051

## Milestone of this PR
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->
`/milestone 1.6.0`

## What type of PR is this

<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
-->
> /kind feature

## Proposed Changes

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests

<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->

Added some Helm chart testing CI tests to validate some behavior.

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added my PR doesn't contain that functionality.
  - [ ] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [ ] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the doc update and the link is:
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

The policies that are flagged as having minimum Kyverno 1.6.0 will produce errors if someone tries to install or upgrade to this version of the Helm chart and the Kyverno deployed is too old.

chipzoller reviewed (changes requested):
@treydock see my updated PR. I missed a couple `match.any` statements when a policy had two rules (I changed the top rule but not the bottom). Make sure you grab them.

treydock commented:
> @treydock see my updated PR. I missed a couple `match.any` statements when a policy had two rules (I changed the top rule but not the bottom). Make sure you grab them.

It looks like I changed things to `match.any` for everything, the match and exclude blocks were pretty much copy+paste between templates as they are identical in most places.

treydock commented:
@chipzoller I added some of the non-PSS policies (so far just non-root groups) to the `other` folder in chart and they will be opt-in via different value than the security standards policies.  Let me know if you think there are other policies I should put into this `other` directory to make easy to deploy with the Helm chart.

chipzoller commented:
Sounds good.

We should probably name the policies that will be installed by default and also opted in with `other`. How can users see that prior to installing? It may also be a good idea to make a note pointing them to https://main.kyverno.io/policies/ because there are not only other policies there, but like I said the host port range policy is an alternate to the one included by default and it can be located from the site.

chipzoller commented:
Also should update some of the following:

```sh
helm show chart kyverno/kyverno-policies
apiVersion: v1
appVersion: v1.5.8
description: Kubernetes Native Policy Management Policies
home: https://kyverno.io/
icon: https://github.com/kyverno/kyverno/raw/main/img/logo.png
keywords:
- kubernetes
- nirmata
- policy agent
- validating webhook
- admissions controller
kubeVersion: '>=1.10.0-0'
maintainers:
- name: Nirmata
  url: https://kyverno.io/
name: kyverno-policies
sources:
- https://github.com/kyverno/kyverno
version: v2.1.10
```

chipzoller commented on charts/kyverno-policies/templates/_helpers.tpl:
```diff
@@ -44,3 +44,35 @@ helm.sh/chart: {{ template "kyverno-policies.chart" . }}
+{{/* Get deployed Kyverno version from Kubernetes */}}
+{{- define "kyverno-policies.kyvernoVersion" -}}
+{{- $version := "" -}}
+{{- with (lookup "apps/v1" "Deployment" .Release.Namespace "kyverno") -}}
```
This will break if the user deploys Kyverno into a Namespace other than `kyverno` yes?

treydock replied:
It uses `.Release.Namespace` so this would not work if the kyverno-policies chart is added to different namespace than Kyverno.  However the version check functions will assume everything is okay if unable to detect the version so if the namespaces don't match, this chart will still install.

chipzoller commented:
And because these PSS policies represent not only a significant rework and addition of new policies but only support 1.6, we probably should bump chart and app versions accordingly.

```sh
$ helm search repo kyverno-policies -l
NAME                            CHART VERSION   APP VERSION     DESCRIPTION
kyverno/kyverno-policies        v2.1.10         v1.5.8          Kubernetes Native Policy Management Policies
kyverno/kyverno-policies        v2.1.7          v1.5.5          Kubernetes Native Policy Management Policies
kyverno/kyverno-policies        v2.1.6          v1.5.4          Kubernetes Native Policy Management Policies
kyverno/kyverno-policies        v2.1.5          v1.5.3          Kubernetes Native Policy Management Policies
kyverno/kyverno-policies        v2.1.4          v1.5.2          Kubernetes Native Policy Management Policies
kyverno/kyverno-policies        v2.1.3          v1.5.1          Kubernetes Native Policy Management Policies
kyverno/kyverno-policies        v2.1.2          v1.5.1          Kubernetes Native Policy Management Policies
kyverno/kyverno-policies        v2.1.1          v1.5.0          Kubernetes Native Policy Management Policies
kyverno/kyverno-policies        v2.1.0          v1.5.0-rc1      Kubernetes Native Policy Management Policies
```

treydock commented:
> And because these PSS policies represent not only a significant rework and addition of new policies but only support 1.6, we probably should bump chart and app versions accordingly.

So this is the issue with opening a PR against main, the versions in main are very old and very out-of-sync with release branches. I could target `release-1.6` and rebase this against that branch, where the versions are newer but then the version check would fail on main branch once merged per my questions in previous comments to @realshuting about the fact main branch tag reports old v1.5.x tag as latest.

treydock commented:
> Sounds good.
> 
> We should probably name the policies that will be installed by default and also opted in with `other`. How can users see that prior to installing? It may also be a good idea to make a note pointing them to https://main.kyverno.io/policies/ because there are not only other policies there, but like I said the host port range policy is an alternate to the one included by default and it can be located from the site.

I think there are a few options, at the very least need to include the list of policies in `README.md` that is specific to the chart.  Could potentially hardcode the list(s) into the NOTES file which prints when the Helm chart is installed.

treydock commented:
> Also should update some of the following:
> 
> ```shell
> helm show chart kyverno/kyverno-policies
> apiVersion: v1
> appVersion: v1.5.8
> description: Kubernetes Native Policy Management Policies
> home: https://kyverno.io/
> icon: https://github.com/kyverno/kyverno/raw/main/img/logo.png
> keywords:
> - kubernetes
> - nirmata
> - policy agent
> - validating webhook
> - admissions controller
> kubeVersion: '>=1.10.0-0'
> maintainers:
> - name: Nirmata
>   url: https://kyverno.io/
> name: kyverno-policies
> sources:
> - https://github.com/kyverno/kyverno
> version: v2.1.10
> ```

Anything in particular needing to be updated?  The versions are a known issue with opening PR against main branch, but wasn't sure if other data there needs to be updated.

chipzoller commented:
> but then the version check would fail on main branch once merged per my questions in previous comments to @realshuting about the fact main branch tag reports old v1.5.x tag as latest.

I talked to Shuting about this before she went on leave, and it's a temporary problem which will eventually get ironed out.

chipzoller commented:
> > Also should update some of the following:
> > ```shell
> > helm show chart kyverno/kyverno-policies
> > apiVersion: v1
> > appVersion: v1.5.8
> > description: Kubernetes Native Policy Management Policies
> > home: https://kyverno.io/
> > icon: https://github.com/kyverno/kyverno/raw/main/img/logo.png
> > keywords:
> > - kubernetes
> > - nirmata
> > - policy agent
> > - validating webhook
> > - admissions controller
> > kubeVersion: '>=1.10.0-0'
> > maintainers:
> > - name: Nirmata
> >   url: https://kyverno.io/
> > name: kyverno-policies
> > sources:
> > - https://github.com/kyverno/kyverno
> > version: v2.1.10
> > ```
> 
> Anything in particular needing to be updated? The versions are a known issue with opening PR against main branch, but wasn't sure if other data there needs to be updated.

The `description` for the `kyverno-policies` chart is really something like "Kubernetes Pod Security Standards implemented as Kyverno policies".

For `keywords`, should `nirmata` still be there or was this a holdover from pre-CNCF donation?

`kubeVersion` should be in alignment with the official support matrix for Kyverno.

`maintainers` ditto as above.

`sources` shouldn't this point to https://github.com/kyverno/policies/tree/main/pod-security or just kyverno/policies?

treydock commented:
> I talked to Shuting about this before she went on leave, and it's a temporary problem which will eventually get ironed out.

So the E2E tests for this PR are failing because of the tag in main, I think merging release-1.6 into main will fix the issue but there are conflicts and not sure if some other better solution.  If she is on leave, someone else we can get to fix this so E2E tests will work?

chipzoller commented:
That might only be @JimBugwadia at this point.

treydock commented:
> The `description` for the `kyverno-policies` chart is really something like "Kubernetes Pod Security Standards implemented as Kyverno policies".

Made this change

> 
> For `keywords`, should `nirmata` still be there or was this a holdover from pre-CNCF donation?

I'm not sure on this. The keywords are also the same for the main kyverno chart too.  Should I delete nirmata from both?
 
> `kubeVersion` should be in alignment with the official support matrix for Kyverno.

What is this for 1.6.x?  The main docs website still looks to only go up to 1.5.x.
 
> `maintainers` ditto as above.

What should new value be?

> `sources` shouldn't this point to https://github.com/kyverno/policies/tree/main/pod-security or just kyverno/policies?

I pointed this at policies repo and homepage at kyverno.io/policies since that's the website for those policies I believe.

chipzoller commented:
I also feel like in the README for this chart, we ought to provide a description of the purpose of the chart and its contents. If you agree, would you like me to write something up and post here as a comment?

treydock commented:
> I also feel like in the README for this chart, we ought to provide a description of the purpose of the chart and its contents. If you agree, would you like me to write something up and post here as a comment?

That sounds fine if you want to write something up.

chipzoller commented:
See https://gist.github.com/chipzoller/51f225a48930d6874eb0b026746a00ca

realshuting commented:
Hi @treydock - sorry for the delay.

We added `1.6-dev` tag to the main branch, and use it for dev-tests. 

From the failure log, I can see 1.6-dev tag was used (because of this change `sed -i -e "s|tag:.*|tag: $(git describe --tag)|" charts/kyverno/values.yaml`),
```
Warning  Failed     3m25s (x4 over 5m1s)  kubelet            Failed to pull image "ghcr.io/kyverno/kyvernopre:1.6-dev-116-g9b9ba3b6": rpc error: code = NotFound desc = failed to pull and unpack image "ghcr.io/kyverno/kyvernopre:1.6-dev-116-g9b9ba3b6": failed to resolve reference "ghcr.io/kyverno/kyvernopre:1.6-dev-116-g9b9ba3b6": ghcr.io/kyverno/kyvernopre:1.6-dev-116-g9b9ba3b6: not found
```
while images were built (by Docker Image Build AMD64) using tag `git describe --match "v[0-9]*"` (make docker-build-all-amd64).

Both tags represent the latest main branch, but `1.6-dev` seems more intuitive. We can change the following Make targets to use `IMAGE_TAG_DEV` tag, i.e., `1.6-dev-116-g9b9ba3b6`:
https://github.com/kyverno/kyverno/blob/9502a8ea8f5427d98806390e0b4fd535aa579623/Makefile#L68-L69
https://github.com/kyverno/kyverno/blob/9502a8ea8f5427d98806390e0b4fd535aa579623/Makefile#L115-L116
https://github.com/kyverno/kyverno/blob/9502a8ea8f5427d98806390e0b4fd535aa579623/Makefile#L157-L158

realshuting commented on .github/workflows/e2e.yaml:
```diff
@@ -114,7 +114,8 @@ jobs:
           sed -i -e "s|fullnameOverride:.*|fullnameOverride: kyverno|g" charts/kyverno/values.yaml
           sed -i -e "s|namespace:.*|namespace: kyverno|g" charts/kyverno/values.yaml
-          sed -i -e "s|tag:  # replaced in e2e tests|tag: $(git describe --match "v[0-9]*")|" charts/kyverno/values.yaml
+          sed -i -e "s|tag:.*|tag: $(git describe --tag)|" charts/kyverno/values.yaml
```
@treydock - this command also replaces tag for busybox container, is that what we want?

treydock replied:
I'm not sure I know what busybox container you're referring to?  When I run the command it replaces tag for main Kyverno container and init container:

```
$ sed -i -e "s|tag:.*|tag: $(git describe --tag)|" charts/kyverno/values.yaml
$ git diff
diff --git a/charts/kyverno/values.yaml b/charts/kyverno/values.yaml
index 696df1a4..63f8e94f 100644
--- a/charts/kyverno/values.yaml
+++ b/charts/kyverno/values.yaml
@@ -16,14 +16,14 @@ rbac:
 image:
   repository: ghcr.io/kyverno/kyverno
   # Defaults to appVersion in Chart.yaml if omitted
-  tag:  # replaced in e2e tests
+  tag: 1.6-dev-127-g79006ec2
   pullPolicy: IfNotPresent
   pullSecrets: []
   # - secretName
 initImage:
   repository: ghcr.io/kyverno/kyvernopre
   # If initImage.tag is missing, defaults to image.tag
-  tag:  # replaced in e2e tests
+  tag: 1.6-dev-127-g79006ec2
   # If initImage.pullPolicy is missing, defaults to image.pullPolicy
   pullPolicy:
   # No pull secrets just for initImage; just add to image.pullSecrets
```

realshuting commented:
I fixed the chart testing for install via #3159, now chart testing for policies failed as it complains the chart version. Let me check how can we bump that in main.

prateekpandey14 commented on charts/kyverno-policies/templates/baseline/disallow-capabilities.yaml:
```diff
@@ -0,0 +1,58 @@
+{{- $name := "disallow-capabilities" }}
+{{- if eq (include "kyverno-policies.podSecurityBaseline" (merge (dict "name" $name) .)) "true" }}
+{{- include "kyverno-policies.supportedKyvernoCheck" (dict "top" . "ver" ">= v1.6.0-rc1") }}
```
Do we need to consider the prerelease version as well , or we can just use v1.6.0 here ?

treydock replied:
That depends entirely on what git tags are exposed. For tests to pass before v1.6.0, must also allow pre-releases otherwise tests won't pass until v1.6.0 tag exists..\

treydock replied:
I re-read docs more carefully, I updated to be `1.6.0-0` which should catch pre-releases.

realshuting commented:
Hi @treydock - I've bumped Kyverno app version and chart version in https://github.com/kyverno/kyverno/pull/3160/files, but  chart testing for policies still failed because of low version:

```
Error: template: kyverno-policies/templates/restricted/restrict-volume-types.yaml:3:4: executing "kyverno-policies/templates/restricted/restrict-volume-types.yaml" at <include "kyverno-policies.supportedKyvernoCheck" (dict "top" . "ver" ">= v1.6.0-rc1")>: error calling include: template: kyverno-policies/templates/_helpers.tpl:75:8: executing "kyverno-policies.supportedKyvernoCheck" at <fail (printf "Kyverno version is too low, expected %s" $supportedKyverno)>: error calling fail: Kyverno version is too low, expected >= v1.6.0-rc1
```

Can you please help here?

treydock commented:
> Hi @treydock - I've bumped Kyverno app version and chart version in https://github.com/kyverno/kyverno/pull/3160/files, but chart testing for policies still failed because of low version:
> 
> ```
> Error: template: kyverno-policies/templates/restricted/restrict-volume-types.yaml:3:4: executing "kyverno-policies/templates/restricted/restrict-volume-types.yaml" at <include "kyverno-policies.supportedKyvernoCheck" (dict "top" . "ver" ">= v1.6.0-rc1")>: error calling include: template: kyverno-policies/templates/_helpers.tpl:75:8: executing "kyverno-policies.supportedKyvernoCheck" at <fail (printf "Kyverno version is too low, expected %s" $supportedKyverno)>: error calling fail: Kyverno version is too low, expected >= v1.6.0-rc1
> ```
> 
> Can you please help here?

So this failure is based on the tag and it was comparing to `v1.6.0-rc1` which might have been seen as breaking with `1.6-dev` so I changed the version comparison to what I hope matches all pre-releases using `1.6.0-0` per Helm docs.

realshuting reviewed (approved):
/lgtm

Thanks @treydock !

prateekpandey14 reviewed (approved):
lgmt, thanks @treydock
